### DISCUSSION TITLE
dp error in high altitude preprocessing

### DISCUSSION THREAD

**Discussion by allayur** [WAM2layers/WAM2layers#190] [Q&A] [answered]
I am running current version of WAM2layers for Tibet source_region 
In preprocessing step I often get error 
assert np.all(dp > 0), "Pressure levels should increase monotonically"
AssertionError: Pressure levels should increase monotonically
I automatically download ERA5 by the script provided
Can it because the surface pressure for the high altitude is low? I run on pressure levels. The last one is 1000

**Comment by Peter9192** [comment, marked as answer]
Hi @allayur, welcome and thanks for reaching out. Like much of the recent developments, the pressure-level functionality hasn't been thoroughly tested yet, especially in mountainous terrain. So it's great to have your feedback!

We built in this superficial check to catch these kind of unanticipated situations. The error originates here:

https://github.com/WAM2layers/WAM2layers/blob/9bc85f741a8742e8131f468bdc5f3bd37047da85/src/wam2layers/preprocessing/era5.py#L170-L172. I'd suggest you have a look at this part of the code to understand what is going on.

Indeed, it could have to do with a low surface pressure at high altitude, or perhaps it has something to do with the value of `p_boundary` which is hardcoded on line 151. Perhaps `p_surface` and `p_boundary` are the same, such that the difference is exactly 0? My first try would be to set a lower value for `p_boundary`. Otherwise, you'd have to carefully examine the values of the data just before this check, to see if you can pinpoint the problem.

Are you able to install a development version of the code, and test a bit with printing some of the values just before reaching this check? See https://wam2layers.readthedocs.io/en/latest/develop.html#developer-s-guide for detailed instructions.

Please let us know if you find out what the problem was, so others may benefit from your scrutiny.

**Comment by allayur** [comment]
Thank you very much @Peter9192 . Lowering p_boundary solved the issue.

**Comment by Peter9192** [comment]
Thanks for reporting back and great that this solved your issue! 

We've been talking about making p_boundary optional, and also about preparing a list of best practices/frequently asked questions/guidance/eperience. Let's keep this issue open until we have that kind of follow-up.
